### PR TITLE
Gulp task to create declaration file

### PR DESCRIPTION
**Pull request by Ledragon**
A new gulp task was created to allow for the generation of a bundled declaration file output in the bundles folder.

Also, some (apparently) missing packages were added to the package.json file.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -29,15 +30,23 @@ function bundle(buildConfig, moduleName, outputFile, outputConfig) {
...
-	streams.push(gulp.src('**/*.js', {cwd: 'platform'}).pipe(gulp.dest('angular2-polyfill/platform')));
-	streams.push(gulp.src('**/*.js', {cwd: 'src'}).pipe(gulp.dest('angular2-polyfill/src')));
+	streams.push(gulp.src('**/*.js', { cwd: 'platform' }).pipe(gulp.dest('angular2-polyfill/platform')));
+	streams.push(gulp.src('**/*.js', { cwd: 'src' }).pipe(gulp.dest('angular2-polyfill/src')));
 
 	return merge(streams);
 });
+
+gulp.task('declarations', () => {
+	var project = ts.createProject('tsconfig.json', { outFile: 'angular2-polyfill.js' });
```
Comment (SamVerschueren) [inline gulpfile.js]: `const project`

File: gulpfile.js [hunk 2]
```diff
@@ -29,15 +30,23 @@ function bundle(buildConfig, moduleName, outputFile, outputConfig) {
...
+	var tsResult = project.src()
```
Comment (SamVerschueren) [inline gulpfile.js]: `const tsResult`

File: gulpfile.js [hunk 3]
```diff
@@ -29,15 +30,23 @@ function bundle(buildConfig, moduleName, outputFile, outputConfig) {
...
+		.pipe(ts(project));
+	return tsResult.dts.pipe(gulp.dest('bundles'))
+
```
Comment (SamVerschueren) [inline gulpfile.js]: remove empty line

File: gulpfile.js [hunk 4]
```diff
@@ -29,15 +30,23 @@ function bundle(buildConfig, moduleName, outputFile, outputConfig) {
...
 		.then(() => del('angular2-polyfill'));
 });
 
 gulp.task('cp', () => {
 	const streams = [];
 	streams.push(gulp.src('*.js').pipe(gulp.dest('angular2-polyfill')));
-	streams.push(gulp.src('**/*.js', {cwd: 'platform'}).pipe(gulp.dest('angular2-polyfill/platform')));
-	streams.push(gulp.src('**/*.js', {cwd: 'src'}).pipe(gulp.dest('angular2-polyfill/src')));
+	streams.push(gulp.src('**/*.js', { cwd: 'platform' }).pipe(gulp.dest('angular2-polyfill/platform')));
+	streams.push(gulp.src('**/*.js', { cwd: 'src' }).pipe(gulp.dest('angular2-polyfill/src')));
```
Comment (SamVerschueren) [inline gulpfile.js]: remove whitespaces

File: gulpfile.js [hunk 5]
```diff
@@ -29,15 +30,23 @@ function bundle(buildConfig, moduleName, outputFile, outputConfig) {
...
+	return bundle(bundleConfig, NG2_POLYFILL_BUNDLE_CONTENT, './bundles/angular2-polyfill.js', { sourceMaps: true })
 		.then(() => del('angular2-polyfill'));
 });
 
 gulp.task('cp', () => {
 	const streams = [];
 	streams.push(gulp.src('*.js').pipe(gulp.dest('angular2-polyfill')));
-	streams.push(gulp.src('**/*.js', {cwd: 'platform'}).pipe(gulp.dest('angular2-polyfill/platform')));
-	streams.push(gulp.src('**/*.js', {cwd: 'src'}).pipe(gulp.dest('angular2-polyfill/src')));
+	streams.push(gulp.src('**/*.js', { cwd: 'platform' }).pipe(gulp.dest('angular2-polyfill/platform')));
```
Comment (SamVerschueren) [inline gulpfile.js]: remove whitespaces

File: gulpfile.js [hunk 6]
```diff
@@ -29,15 +30,23 @@ function bundle(buildConfig, moduleName, outputFile, outputConfig) {
 }
 
 gulp.task('bundle', ['cp'], () => {
-	return bundle(bundleConfig, NG2_POLYFILL_BUNDLE_CONTENT, './bundles/angular2-polyfill.js', {sourceMaps: true})
+	return bundle(bundleConfig, NG2_POLYFILL_BUNDLE_CONTENT, './bundles/angular2-polyfill.js', { sourceMaps: true })
```
Comment (SamVerschueren) [inline gulpfile.js]: remove whitespaces

**Comment by SamVerschueren** [comment]
Nice! I think this is what I am looking for! Have to reconfirm tomorrow but looks promising. Just added a few remarks.

> Also, some (apparently) missing packages were added to the package.json file.

Good catch, something went wrong with my latest push :)

**Comment by Ledragon** [comment]
I made the corrections according to your comments. Would you like a linter to be added? If so, which one (jslint/eslint)? What about TypeScript linting? Maybe those questions should be adressed in another issue...

**Comment by SamVerschueren** [comment]
Just tested the type definition. It works but the module names should be prefixed with `angular2-polyfill`. This was an issue I had with systemjs as well and not sure how I can resolve this easily, so feedback is more then welcome.

[Angular2](https://github.com/angular/angular/tree/master/modules) actually has one repository for all the angular core stuff if I'm not mistaken. (angular2, angular2-dart). Every project has it's own `package.json` file and at the root of the angular repository there is also a `package.json`. The root package contains all the dependencies regarding building and gulpfile stuff. The angular package.json files contain only the dependencies necessary for that piece. This makes it possible to actually build my bundle without copying everything to a subdirectory and would make generating a bundle d.ts file a lot easier as well.

```
package.json
gulpfile.js
angular2-polyfill
    - core.ts
    - router.ts
    - src
        ...
```

When releasing, the `angular2-polyfill` directory should be published to npm and not the root. @Ledragon  any experience with this? Does it looks reasonable?

**Comment by Ledragon** [comment]
@SamVerschueren That would solve the problem of relative paths at least, and it would make sense to only deploy the `angular2-polyfill` directory. I don't really have any experience about it, but it sounds ok. Plus, if the angular team does it, why not? :p 
I researched, but could not find a better way of doing it as of now.

**Comment by SamVerschueren** [comment]
Might have to revert the merge. Can't seem to figure out how I can prefix the modules with `angular2-polyfill`. It always compiles to `core` instead of `angular2-polyfill/core`.

**Comment by SamVerschueren** [comment]
@Ledragon I created a separate branch with a refactored code structure called [typedefs](https://github.com/SamVerschueren/angular2-polyfill/tree/typedefs). I managed to get things working regarding the bundling but the type definitions were still not correct. Can't seem to figure out how I have to configure them properly. Care to take a look?

**Comment by Ledragon** [comment]
Ok, just got to merge your updated project in my fork (not exactly a GitHub expert), then I'm on it!

**Comment by SamVerschueren** [comment]
Thanks for looking into this! Really appreciate it.

This might help: https://help.github.com/articles/syncing-a-fork/

**Comment by Ledragon** [comment]
Yep, managed to get it with the new branch. Time to start on the actual stuff!

**Comment by Ledragon** [comment]
Man... That was hot! I will submit a new pull request on the branch with something that works (hopefully) as expected.

**Pull request merged**